src/ts_catalog/continuous_aggs_watermark.c

```diff
@@ -384,6 +387,19 @@ cagg_watermark_update_internal(int32 mat_hypertable_id, int64 new_watermark, boo
 				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
 				 errmsg("watermark not defined for continuous aggregate: %d", mat_hypertable_id)));
 	}
+
+	/*
+	 * During query planning, the values of the watermark function are constified using the
+	 * constify_cagg_watermark() function. However, this function's value changes when we update the
+	 * Cagg (the volatility of the function is STABLE not IMMUTABLE). To ensure that caches, such as
+	 * the query plan cache, are properly evicted, we send an invalidation message for the
+	 * hypertable.
+	 */
+	if (invalidate_rel_cache)
+	{
+		DEBUG_WAITPOINT("cagg_watermark_update_internal_before_refresh");
+		CacheInvalidateRelcacheByRelid(ht_relid);
+	}
```

Comment:
I guess this logic should be moved to `cagg_watermark_update_scan_internal` at the point that we *really* update the catalog table. WDYT?

Reply:
Moved as discussed